### PR TITLE
Og check field cardinality

### PR DESCRIPTION
**Pull request by damiankloip**
This just ports the `og_check_field_cardinality` function, so should be pretty self sufficient.

I created a new `OgHelper` class, as I think we should start using this for things that are more helpers than API. The `Og` class will get way overloaded (more than it is already :)).

### CODE REVIEW

**Comment by damiankloip** [comment]
ping @amitaibu

File: src/OgHelper.php [hunk 1]
```diff
@@ -0,0 +1,45 @@
...
+namespace Drupal\og;
+
+use Drupal\Core\Entity\ContentEntityInterface;
+use Drupal\Core\Field\FieldException;
+use Drupal\Core\Field\FieldStorageDefinitionInterface;
+
+/**
+ * OG helper methods.
+ */
+class OgHelper {
```
Comment (amitaibu) [inline src/OgHelper.php]: maybe this should be called `OgGroupAudience` or `OgGroupAudienceHelper`? - I think it would make its content clearer

File: src/OgHelper.php [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+
+  /**
+   * Return TRUE if a field can be used and has not reached maximum values.
+   *
+   * @param \Drupal\Core\Entity\ContentEntityInterface $entity
+   *   The content entity to check the field cardinality for.
+   * @param string $field_name
+   *   The field name to check the cardinality of.
+   *
+   * @return bool
```
Comment (amitaibu) [inline src/OgHelper.php]: I think should also include a `@throws` in the PHP docs.

File: src/OgHelper.php [hunk 3]
```diff
@@ -0,0 +1,45 @@
...
+   */
+  public static function checkFieldCardinality(ContentEntityInterface $entity, $field_name) {
+    $field_definition = $entity->getFieldDefinition($field_name);
+
+    if (!$field_definition) {
```
Comment (amitaibu) [inline src/OgHelper.php]: we should probably also check `Og::isGroupAudienceField` (which could move here, and maybe be called `OgAudienceField::isValid`)

File: src/OgHelper.php [hunk 4]
```diff
@@ -0,0 +1,45 @@
...
+   * Return TRUE if a field can be used and has not reached maximum values.
+   *
+   * @param \Drupal\Core\Entity\ContentEntityInterface $entity
+   *   The content entity to check the field cardinality for.
+   * @param string $field_name
+   *   The field name to check the cardinality of.
+   *
+   * @return bool
+   */
+  public static function checkFieldCardinality(ContentEntityInterface $entity, $field_name) {
```
Comment (amitaibu) [inline src/OgHelper.php]: I'm still trying to think what would be better `EntityInterface` or `ContentEntityInterface`. For example in Og::userAccess() we are checking `EntityInterface` and if not a `ContentEntityInterface` we return early.

I usually tend to prefer that API functions would be strict, and throw exceptions if they don't get the right data.

So the question is - should we change other places to use `ContentEntityInterface` as-well?

File: tests/src/Unit/CheckFieldCardinalityTest.php [hunk 5]
```diff
@@ -0,0 +1,110 @@
...
+   *
+   * @expectedException \Drupal\Core\Field\FieldException
+   */
+  public function testFieldCardinalityNoDefinition() {
+    $field_name = 'test_field_no_definition';
+    $entity_prophecy = $this->prophesize(ContentEntityInterface::class);
+
+    $entity_prophecy->getFieldDefinition($field_name)
+      ->willReturn(NULL);
+    // The bundle() and getEntityTypeId() methods will be called for the
```
Comment (amitaibu) [inline tests/src/Unit/CheckFieldCardinalityTest.php:65]: nit: Line break above.

File: src/OgHelper.php (hunk 4 above)
Comment (damiankloip) [inline src/OgHelper.php]: I think it depends on the usage tbh. Some function you will want to be able to call on more entities, some do not need that so much, like this one. We use audience field checking quite broadly, so having to check the interface before calling it in all those places is not ideal.

We need to type hint `ContentEntityInterface` or `FieldableEntityInterface` (which might be a better choice) here as you have no methods available for getting field definitions etc..

File: src/OgHelper.php (hunk 1 above)
Comment (damiankloip) [inline src/OgHelper.php]: Sure, we can be more granular. I would go for `OgGroupAudienceHelper`.

File: src/OgHelper.php (hunk 4 above)
Comment (amitaibu) [inline src/OgHelper.php]: What would be the (real world) use case for `FieldableEntityInterface`?

Anyway, just wanted to put that question out there - we can revisit it later

File: src/OgHelper.php (hunk 1 above)
Comment (amitaibu) [inline src/OgHelper.php]: :+1:

File: src/OgHelper.php (hunk 4 above)
Comment (damiankloip) [inline src/OgHelper.php]: Not sure really, surprise surprise - the only thing using it is `ContentEntityInterface`. Let's just leave it like this for now then.

Comment (amitaibu) [inline src/OgHelper.php]: I agree. A good practice I find is always trying to find a _real_ use case - usually saves a lot of complexity ;)

File: src/OgHelper.php (hunk 3 above)
Comment (damiankloip) [inline src/OgHelper.php]: I like that idea, but I think we might want to move some other things over too maybe like maybe Og::`getAllGroupAudienceFields`. So let's do a separate PR for this move?

Comment (amitaibu) [inline src/OgHelper.php]: sure.

Comment (amitaibu) [inline src/OgHelper.php]: but the `Og::isGroupAudienceField` would be better added in this PR I think.

**Comment by damiankloip** [comment]
@amitaibu updated all the things.

File: src/OgGroupAudienceHelper.php [hunk 6]
```diff
@@ -0,0 +1,51 @@
...
+ */
+
+namespace Drupal\og;
+
+use Drupal\Core\Entity\ContentEntityInterface;
+use Drupal\Core\Field\FieldException;
+use Drupal\Core\Field\FieldStorageDefinitionInterface;
+
+/**
+ * OG audience helper methods.
```
Comment (amitaibu) [inline src/OgGroupAudienceHelper.php]: `OG audience field helper methods.`

File: src/OgGroupAudienceHelper.php [hunk 7]
```diff
@@ -0,0 +1,51 @@
...
+   *
+   * @return bool
+   *
+   * @throws \Drupal\Core\Field\FieldException
+   */
+  public static function checkFieldCardinality(ContentEntityInterface $entity, $field_name) {
+    $field_definition = $entity->getFieldDefinition($field_name);
+
+    if (!$field_definition) {
+      throw new FieldException(sprintf('No "%s" field found for %s %s entity', $field_name, $entity->bundle(), $entity->getEntityTypeId()));
```
Comment (amitaibu) [inline src/OgGroupAudienceHelper.php:35]: in some other places we use `new FormattableMarkup` instead of `sprintf` which makes the string more readable IMO

File: src/OgGroupAudienceHelper.php [hunk 8]
```diff
@@ -0,0 +1,51 @@
...
+    }
+
+    if (!Og::isGroupAudienceField($field_definition)) {
+      throw new FieldException(sprintf('"%s" field on %s %s entity is not an audience field.', $field_name, $entity->bundle(), $entity->getEntityTypeId()));
```
Comment (amitaibu) [inline src/OgGroupAudienceHelper.php:39]: are we throwing the right exception here? (if not, could be a follow up)

File: tests/src/Unit/CheckFieldCardinalityTest.php [hunk 9]
```diff
@@ -0,0 +1,143 @@
...
+    OgGroupAudienceHelper::checkFieldCardinality($entity_prophecy->reveal(), $field_name);
+  }
+
+  /**
+   * @covers ::checkFieldCardinality
+   *
+   * @expectedException \Drupal\Core\Field\FieldException
+   */
+  public function testFieldCardinalityNotAudienceField() {
+    $field_name = 'test_field_no_definition';
```
Comment (amitaibu) [inline tests/src/Unit/CheckFieldCardinalityTest.php]: `test_field_not_audience`

Comment (amitaibu) [inline tests/src/Unit/CheckFieldCardinalityTest.php]: or even `test_field_not_og_audience`

File: tests/src/Unit/CheckFieldCardinalityTest.php [hunk 10]
```diff
@@ -0,0 +1,143 @@
...
+    $entity_prophecy = $this->prophesize(ContentEntityInterface::class);
+
+    $field_definition_prophecy = $this->prophesize(FieldDefinitionInterface::class);
+    $field_definition_prophecy->getType()
+      ->willReturn('invalid_field_type')
+      ->shouldBeCalled();
+    $field_definition_prophecy->getFieldStorageDefinition()
```
Comment (amitaibu) [inline tests/src/Unit/CheckFieldCardinalityTest.php]: why do we need this?

File: tests/src/Unit/CheckFieldCardinalityTest.php [hunk 11]
```diff
@@ -0,0 +1,143 @@
...
+        ->shouldBeCalled();
+    }
+
+    $this->assertSame(OgGroupAudienceHelper::checkFieldCardinality($entity_prophecy->reveal(), $field_name), $expected);
+  }
+
+  /**
+   * Data provider for testFieldCardinality.
+   *
+   * @return array
```
Comment (amitaibu) [inline tests/src/Unit/CheckFieldCardinalityTest.php:125]: I'd list down the values its returning, as it makes it easier to read, rather then go back to `testFieldCardinality`.  maybe something like:

``` php
@return array
  The values to test which correspond to:
  - The count of existing populated items in the field.
  - Field cardinality.
  - The expected result where TRUE signifies the field may be populated by another value. 
```

File: src/OgGroupAudienceHelper.php (hunk 7 above)
Comment (damiankloip) [inline src/OgGroupAudienceHelper.php:35]: ok, sure, I don't mind too much. FormattableMarkup doesn't have any dependencies on the container etc... so seems legit.

File: src/OgGroupAudienceHelper.php (hunk 8 above)
Comment (damiankloip) [inline src/OgGroupAudienceHelper.php:39]: Seems like a field related exception. I think we should add OG specific exceptions though - think I mentioned that somewhere before.

File: tests/src/Unit/CheckFieldCardinalityTest.php (hunk 9 above)
Comment (damiankloip) [inline tests/src/Unit/CheckFieldCardinalityTest.php]: ha, proper nit :D

File: src/OgGroupAudienceHelper.php (hunk 8 above)
Comment (amitaibu) [inline src/OgGroupAudienceHelper.php:39]: indeed you have :) this could be a follow up

File: tests/src/Unit/CheckFieldCardinalityTest.php (hunk 10 above)
Comment (damiankloip) [inline tests/src/Unit/CheckFieldCardinalityTest.php]: Which part? checking getFieldStorageDefinition is not called?

File: src/OgGroupAudienceHelper.php (hunk 7 above)
Comment (damiankloip) [inline src/OgGroupAudienceHelper.php:35]: The reason I don't like FormattablMarkup is becuase it's more a part of the render system. This is not really anything to do with that...

File: src/OgGroupAudienceHelper.php (hunk 6 above)
Comment (damiankloip) [inline src/OgGroupAudienceHelper.php]: :+1:

File: tests/src/Unit/CheckFieldCardinalityTest.php (hunk 10 above)
Comment (amitaibu) [inline tests/src/Unit/CheckFieldCardinalityTest.php]: yes

File: src/OgGroupAudienceHelper.php (hunk 7 above)
Comment (amitaibu) [inline src/OgGroupAudienceHelper.php:35]: I see your point, but in terms of readability it easier to follow a [sentence](https://github.com/amitaibu/og/blob/8.x-1.x/src/Og.php#L408) rather than mentally replacing the `%s` with an argument.

File: src/OgGroupAudienceHelper.php (hunk 8 above)
Comment (damiankloip) [inline src/OgGroupAudienceHelper.php:39]: Yes, let's do an awesome Og Exception follow up.

File: tests/src/Unit/CheckFieldCardinalityTest.php (hunk 10 above)
Comment (damiankloip) [inline tests/src/Unit/CheckFieldCardinalityTest.php]: Could go. I'll remove it.

File: src/OgGroupAudienceHelper.php (hunk 7 above)
Comment (damiankloip) [inline src/OgGroupAudienceHelper.php:35]: I did see that, but guess where the only place that does that is... OG :) Nowhere in core does that pattern. How about just the variables in place in the string??

Comment (amitaibu) [inline src/OgGroupAudienceHelper.php:35]: Well, I saw it in Message - but that's us again ;)

Ok, following core wins!

Comment (amitaibu) [inline src/OgGroupAudienceHelper.php:35]: I've opened issue https://github.com/amitaibu/og/issues/57

Comment (damiankloip) [inline src/OgGroupAudienceHelper.php:35]: haha, fair.

File: tests/src/Unit/CheckFieldCardinalityTest.php (hunk 11 above)
Comment (damiankloip) [inline tests/src/Unit/CheckFieldCardinalityTest.php:125]: Sure, let me add something now.

Comment (damiankloip) [inline tests/src/Unit/CheckFieldCardinalityTest.php:125]: @amitaibu I think this is looking swell now.

**Comment by amitaibu** [comment]
Thanks!

**Comment by damiankloip** [comment]
Cheers! Can now be used in https://github.com/amitaibu/og/pull/40
